crates/subspace-archiving/src/archiver/incremental_record_commitments.rs

```diff
@@ -0,0 +1,194 @@
+
+            self.update_hashing_state(remaining_record_bytes);
+
+            if remaining_record_bytes.len() == bytes_until_full_record {
+                self.finish_hash();
+            }
+
+            bytes
+        };
+
+        // Continue processing records (full and partial) from remaining data
+        bytes.chunks(self.record_size).for_each(|record| {
+            self.update_hashing_state(record);
+
+            // Store hashes of full records
+            if record.len() == self.record_size {
+                self.finish_hash();
+            }
+        });
+    }
+}
+
+impl<'a> IncrementalRecordCommitmentsProcessor<'a> {
+    fn new(
+        incremental_record_commitments: &'a mut IncrementalRecordCommitmentsState,
+        record_size: usize,
+        full: bool,
+    ) -> Self {
+        Self {
+            processed: 0,
+            incremental_record_commitments,
+            record_size,
+            full,
+            hashing_state: Blake2b::<U32>::default(),
+        }
+    }
+
+    /// How many bytes to read before the start of the next record
+    fn bytes_until_full_record(&self) -> usize {
+        let bytes_to_next_record =
+            ((self.processed / self.record_size) + 1) * self.record_size - self.processed;
+        bytes_to_next_record % self.record_size
+    }
+
+    /// Whether commitment for current record needs to be created
+    fn should_commit_to_record(&self, record_position: usize) -> bool {
+        if self.full {
+            // For full segment we need to create the first record and any that are not present yet
+            record_position == 0
+                || self
+                    .incremental_record_commitments
+                    .state
+                    .get(record_position)
+                    .is_none()
+        } else {
+            // For partial segment we need to skip the first record (it is not final) and generate
+            // any other that are currently missing
+            record_position
+                .checked_sub(1)
+                .map(|shifted_position| {
+                    self.incremental_record_commitments
+                        .state
+                        .get(shifted_position)
+                        .is_none()
+                })
+                .unwrap_or_default()
+        }
+    }
+
+    /// In case commitment is necessary for currently processed record, internal hashing state will
+    /// be updated with provided bytes.
+    fn update_hashing_state(&mut self, bytes: &[u8]) {
+        if self.should_commit_to_record(self.processed / self.record_size) {
+            self.hashing_state.update(bytes);
+        }
+        self.processed += bytes.len();
+    }
+
+    /// In case commitment is necessary for currently processed record, internal hashing state will
+    /// be finalized and commitment will be stored in shared state.
+    fn finish_hash(&mut self) {
+        if self.should_commit_to_record(self.processed / self.record_size - 1) {
```

Comment:
`((self.processed / self.record_size) + 1) * self.record_size - self.processed` seems to be always greater than zero. Is there a proof for `self.processed / self.record_size - 1` ? Can we be sure that `self.processed` is always greater than `self.record_size`?

Reply:
Yes, even though it is not that obvious. By the time we call `finish_hash` we have a multiple of `self.record_size` in `self.processed`, which implies that division of one over another will be at least `1`.